provider/materialization_adapter.go

```diff
@@ -0,0 +1,325 @@
+		Name:       types.ColumnName(schema.Entity),
+		NativeType: types.NativeType("string"), // Adjust based on your actual types
+		Type:       types.String,               // Adjust based on your actual types
+	}
+
+	// Create feature column schemas
+	featureCols := make([]types.FeatureCol, 1) // Adjust if you have multiple features
+	featureCols[0] = types.FeatureCol{
+		FeatureCol: types.ColumnSchema{
+			Name:       types.ColumnName(schema.Value),
+			NativeType: types.NativeType("string"), // Adjust based on your actual types
+			Type:       types.String,               // Adjust based on your actual types
+		},
+	}
+
+	// Add timestamp column if available
+	if schema.TS != "" {
+		featureCols[0].TimestampCol = types.ColumnSchema{
+			Name:       types.ColumnName(schema.TS),
+			NativeType: types.NativeType("timestamp"),
+			Type:       types.Timestamp,
+		}
+	}
+
+	// For cases with EntityMappings, you'll need to expand this logic
+	if len(schema.EntityMappings.Mappings) > 0 {
+		// Handle entity mappings based on your requirements
+		// This is a simplification - adjust based on your actual semantics
+		// for entity mappings
+	}
+
+	return types.FeaturesSchema{
+		EntityCol:   entityCol,
+		FeatureCols: featureCols,
+	}
+}
+
+// extractFieldsFromFeatureSchema extracts column schemas from a feature schema
+func extractFieldsFromFeatureSchema(featureSchema types.FeaturesSchema) []types.ColumnSchema {
+	fields := make([]types.ColumnSchema, 0)
+
+	// Add entity column
+	fields = append(fields, featureSchema.EntityCol)
+
+	// Add feature columns
+	for _, featureCol := range featureSchema.FeatureCols {
+		fields = append(fields, featureCol.FeatureCol)
+		// Add timestamp column if it exists
+		if featureCol.TimestampCol.Name != "" {
+			fields = append(fields, featureCol.TimestampCol)
+		}
+	}
+
+	return fields
+}
+
+// convertResourceRecordToRow converts a ResourceRecord to a Row
+func convertResourceRecordToRow(record ResourceRecord) types.Row {
+	// Create a row with three values: entity, value, and timestamp
+	values := make(types.Row, 3)
+
+	// First value is always the entity
+	values[0] = types.Value{
+		Value: record.Entity,
+	}
+
+	// Second value is always the feature value
+	values[1] = types.Value{
+		Value: record.Value,
+	}
+
+	// Third value is the timestamp (or nil if timestamp is zero)
+	if !record.TS.IsZero() {
+		values[2] = types.Value{
+			Value: record.TS,
+		}
+	} else {
+		values[2] = types.Value{
+			Value: nil,
+		}
+	}
+
+	return values
+}
+
+func RowToResourceRecord(row types.Row) (ResourceRecord, error) {
+	rec := ResourceRecord{}
+
+	// Check if we have at least entity and value fields
+	if len(row) < 2 {
+		return rec, fmt.Errorf("row has insufficient columns: expected at least 2, got %d", len(row))
+	}
+
+	// Extract entity - should be first column and a string
+	if row[0].Value == nil {
+		return rec, fmt.Errorf("entity column has nil value")
+	}
+
+	entityVal, ok := row[0].Value.(string)
```

Comment:
Can't we validate some of this stuff with a Dataset? I guess this is a legacy adapter